Doc/library/tempfile.rst

```diff
@@ -75,20 +75,52 @@ The module defines the following user-callable items:
       Added *errors* parameter.
 
 
-.. function:: NamedTemporaryFile(mode='w+b', buffering=-1, encoding=None, newline=None, suffix=None, prefix=None, dir=None, delete=True, *, errors=None)
-
-   This function operates exactly as :func:`TemporaryFile` does, except that
-   the file is guaranteed to have a visible name in the file system (on
-   Unix, the directory entry is not unlinked).  That name can be retrieved
-   from the :attr:`name` attribute of the returned
-   file-like object.  Whether the name can be
-   used to open the file a second time, while the named temporary file is
-   still open, varies across platforms (it can be so used on Unix; it cannot
-   on Windows).  If *delete* is true (the default), the file is
-   deleted as soon as it is closed.
-   The returned object is always a file-like object whose :attr:`!file`
-   attribute is the underlying true file object. This file-like object can
-   be used in a :keyword:`with` statement, just like a normal file.
+.. function:: NamedTemporaryFile(mode='w+b', buffering=-1, encoding=None, newline=None, suffix=None, prefix=None, dir=None, delete=True, *, errors=None, delete_on_close=True)
+
+   This function operates exactly as :func:`TemporaryFile` does, except the
+   following differences:
+
+   * The file is guaranteed to have a visible name in the file system
+     (on Unix, the directory entry is not unlinked).
+   * There is more granularity in the deletion behaviour of the file
+     (see *delete_on_close* below)
+
+   The returned object is always a :term:`file-like object` whose :attr:`!file`
+   attribute is the underlying true file object. This :term:`file-like object` can be
+   used in a :keyword:`with` statement, just like a normal file.  The name of the
+   temporary file can be retrieved from the :attr:`name` attribute of the
+   returned file-like object.
+
+   If *delete* is true (the default) and *delete_on_close* is true (the
+   default), the file is deleted as soon as it is closed. If *delete* is true
+   and *delete_on_close* is false, the file is deleted on context manager exit
+   only, if no context manager was used, then the file is deleted only when the
+   :term:`file-like object` is finalized and hence deletion is not always
+   guaranteed in this case (see :meth:`object.__del__`). If *delete* is false,
+   the value of *delete_on_close* is ignored.
+
+   While the named temporary file is open, the file can always be opened again
+   on POSIX. On Windows, it can be opened again if *delete* is false, or if
+   *delete_on_close* is false, or if the additional open shares delete access
+   (e.g. by calling :func:`os.open` with the flag ``O_TEMPORARY``).  On
+   Windows, if *delete* is true and *delete_on_close* is false, additional
+   opens that do not share delete access (e.g. via builtin :func:`open`) must
+   be closed before exiting the context manager, else the :func:`os.unlink`
+   call on context manager exit will fail with a :exc:`PermissionError`.
```

Comment:
Can we reframe this as:

* the file can be opened again using its name
* on Windows, subsequent opens should share delete access (e.g.)
* however, when `delete` or `delete_on_close` are false, sharing delete access is not necessary, but you should ensure the file is closed before leaving the context manager

This sets it up more positively for the user - what _should_ they do, as opposed to what could go wrong.

Reply:
Please check how I have written this in the new version. Not sure everybody would agree, but I have now written this information the way I would want it to be explained for me without the risk of twisting my brains.

**Additional question:** does it make sense to  write unit tests to cover all these combinations of opening the file the second time, while it is already open.  I have tested this for myself (on Windows) to make sure it is all correct, but I am not sure that it shall go to unit tests.